R/utils.R

```diff
@@ -259,82 +259,500 @@ rowcol_to_dim <- function(row, col) {
+    "row_names" = !is.null(args$x) & isTRUE(args$row_names) & is.null(args$cols), # because default is FALSE
+    "data" = TRUE
+  )
+  default_select <- if (isFALSE(args$col_names) || !is.null(args$rows) || !is.null(args$cols)) {
+    "data"
+  } else {
+    "x"
+  }
+  select <- select %||% default_select
+  valid_cases_choices <- names(valid_cases)
+  match.arg_wrapper(select, choices = valid_cases_choices, fn_name = "wb_dims", several.ok = FALSE)
+
+  if (isFALSE(valid_cases[[select]])) {
+    stop(
+      "You provided a bad value to `select` in `wb_dims()`.\n ",
+      "Please review. see `?wb_dims`.",
+      call. = FALSE
+    )
+  }
+  select
+}
+# it is a wrapper around base::match.arg(), but it doesn't allow partial matching.
+# It also provides a more informative error message in case it fails.
+match.arg_wrapper <- function(arg, choices, several.ok = FALSE, fn_name = NULL) {
+  # Check valid argument names
+  # partial matching accepted
+  fn_name <- fn_name %||% "fn_name"
+  # match.arg(arg, choices = choices, several.ok = several.ok)
+  # Using rlang::arg_match() would remove that.
+  if (!several.ok) {
+    if (length(arg) != 1) {
+      stop(
+        "Must provide a single argument found in ", fn_name, ": ",
+        invalid_arg_nams, "\n", "Use one of ", valid_arg_nams,
+        call. = FALSE
+      )
+    }
+  }
+
+  invalid_args <- !arg %in% choices
+  if (any(invalid_args)) {
+    invalid_arg_nams <- paste0("`", arg[invalid_args], "`", collapse = ", ")
+    multi <- length(invalid_arg_nams) > 0
+    plural_sentence <- ifelse(multi, " is an invalid argument for ", " are invalid arguments for ")
+
+    valid_arg_nams <- paste0("'", choices[choices != ""], "'", collapse = ", ")
+    stop(
+      invalid_arg_nams, plural_sentence, fn_name, ": ", "\n", "Use any of ", valid_arg_nams,
+      call. = FALSE
+    )
+  }
+}
+
+
+#' Helper to specify the `dims` argument
+#'
+#' @description
+#'
+#' `wb_dims()` is experimental, any use case outside the documented ones may work,
+#' but is likely to fail or change.
+#'
+#' `wb_dims()` can be used to help provide the `dims` argument, in the `wb_add_*` functions.
+#' It returns a A1 spreadsheet range ("A1:B1" or "A2").
+#' It can be very useful as you can specify many parameters that interact together
+#' In general, you must provide named arguments. `wb_dims()` will only accept unnamed arguments
+#' if they are `rows`, `cols`, for example `wb_dims(1:4, 1:2)`, that will return "A1:B4".
+#'
+#' `wb_dims()` can also be used with an object (a `data.frame` or a `matrix` for example.)
+#' All parameters are numeric unless stated otherwise.
+#'
+#' # Using `wb_dims()` without an `x` object
+#'
+#' * `rows` / `cols` (if you want to specify a single one, use `from_row` / `from_col`)
+#' * `from_row` / `from_col` the starting position of the `dims`
+#'    (similar to `start_row` / `start_col`, but with a clearer name.)
```

Comment:
Also the intention is different at least in the way I see it. With `start_` you are reading from a column downwards (all 1M rows or 10k rows). With `from_` you are placing a cell range to a specific part of the worksheet.

Reply:
Ok, do you have places where this is implemented elsewhere?

I used your initial ideas from `wb_dims()`, but maybe we can continue the discussion on the issue?

Reply:
Sure. Not sure what you mean with elsewhere. The initial proposal was a quick draft to have something nicer than `rowcol_to_dims()`. I most likely haven't thought about all corner stones. I usually like to get things rolling and fine-tuning can happen afterwards.

Reply:
I mean in other functions of openxlsx2.

But I'll be happy to do modifications afterwards too!

Reply:
Ah, got it. We have `start_` in `wb_to_df()` and it should (soon) be deprecated in all other functions in favor of `dims`. Previously we had competing arguments for selections like `dims`, `cols/rows`, `col/row` and `startCol/startRow`. Now we have only one and soon a nice wrapper for `dims` that can provide the other variants.

We still have remaining `start_` arguments in `wb_add_data()` and its derivatives. These can be removed, but we can do that after the merge. This way we'll lose another bunch of confusing arguments. But there we have to be careful, might require some under the hood conversions, because it will be a last minute change under the hood. We can discuss this somewhere else.